### ISSUE TITLE
Invert Y axis option

### ISSUE THREAD

**Issue by Beastifus** [mapeditor/tiled#249]
Currently the tiling y-axis starts with the height at 0 at the top and incrementing as it goes down the map.   I think it would a be a nice feature, but relatively simple task, to add some kind of checkbox option to invert the Y axis so that (0,0) starts at the bottom and then (0, 1) is the tile above.   

Thanks!

**Comment by jrmgx** [comment]
I'm much in favor of this feature request! :)

**Comment by bjorn** [comment]
Also requested [on the forum](http://discourse.mapeditor.org/t/change-tiled-coordinate-system-to-left-bottom/1704).

**Comment by ShipshapeDetour** [comment]
This feature would be very helpful.

**Comment by bjorn** [comment]
> I think this is a good idea, rather than having to adjust the code in game it would make more sense just to have an option here. Then you wouldn't need to make boilerplate code just to invert the access in game for every object.

@davos555 I realize I'm a bit late with this reply, but I just thought I'd point out that I see this option as only a display option in Tiled. It would not affect the actual coordinates stored in a TMX file. So if you want to use the inverted coordinates in your game, you'll still have to invert them while or after loading the map. That is because if the TMX format would support storing inverted coordinates, then every loader out there would suddenly need to add such conversion code.

**Comment by kibertoad** [comment]
@bjorn Why can't this conversion be applied transparently during the save process so that loaders do not have to do anything at all?.. Aren't coords actually saved?

**Comment by ShipshapeDetour** [comment]
That's a sound idea, but the most important part for me is changing the y
coordinate in the status bar. It's much easier to implement.

**Comment by bjorn** [comment]
> @bjorn Why can't this conversion be applied transparently during the save process so that loaders do not have to do anything at all?.. Aren't coords actually saved?

@kibertoad There are many loaders either integrated into various game development frameworks or provided by 3rd parties, that both read the map file and render the map. When an option to invert the Y axis in the TMX file is introduced, then in order to render the map correctly, they would all need to be aware of this option and invert the Y axis when appropriate while rendering.

I think it is better to keep a particular coordinate system on the file format level so that we are all on the same page. At the same time, I understand that some people are working with their coordinates inverted, and that it could help if Tiled could display them as such.

@ShipshapeDetour In addition to the status bar I think it'll need to affect the Properties view as well, but still you're right that this is less work than changing the interpretation of the data.

**Comment by kibertoad** [comment]
@bjorn Yes, if this is implemented as "invertCoords = true" parameter. What I mean by "transparently" is that if "invertCoords = true" is set, then on saving map Tiled will replace "X=1, Y=1" with "X=1, Y=100", and on loading (through Tiled) swap it all back.

**Comment by viking** [comment]
Any news on this feature? Would the option also affect x and y coordinates in Python export plugins?

**Comment by bjorn** [comment]
@viking This is currently not one of my priorities, though I think it would be a good issue for somebody new to Tiled development to look into, so I've now tagged it as such.

To avoid making the map format more complex, it is planned to only affect the display of coordinates in Tiled. So it would not affect Python export plugins.

**Comment by divitto** [comment]
I'm interested in trying to implement this feature. Just to be clear all that needs to be done is add a check box in the settings to flip the y coordinates of tiles that is being displayed in the status bar right? Nothing else involving actually saving the maps or modifying the properties view as mentioned above? I'm confident i can do that, but this is my first time contributing so any pointers would be greatly apreciated. :)

**Comment by bjorn** [comment]
> Just to be clear all that needs to be done is add a check box in the settings to flip the y coordinates of tiles that is being displayed in the status bar right? Nothing else involving actually saving the maps or modifying the properties view as mentioned above?

Almost. Note that the status bar is not the only place where coordinates are displayed. And it does also affect the Properties view, since the flipping should probably be applied to both tile coordinates and the object coordinates:

* Object coordinates are also displayed (optionally) in the Objects view. It's a column you can enable by right-clicking on the header. So the flipping needs to be supported also by the `MapObjectModel`.

* Object coordinates are editable in the Properties view. In this case, the flipping needs to happen before setting the value of the property and when applying changes the user made using this control (in reverse). This is the `PropertyBrowser`.

Thanks for working on this, @divitto, and feel free to ask any additional questions!

**Comment by divitto** [comment]
I've been looking over the source code in my spare time the last few days. I think it would help me a lot if you could more clearly define what exactly needs to be done, maybe a high level list of things you would do. I'm having a hard time figuring out where to start so to speak, and i think it stems from not clearly understanding the goal and everything it affects. Thanks for your time i understand you are busy with the real problems. :)

**Comment by bjorn** [comment]
@divitto So, there are several places that show a tile location to the user. One such location is the hovered coordinate in the status bar. Here's one place where this text is created:

https://github.com/bjorn/tiled/blob/da9ba8f0c7138850b7c2b9b23e6544d0705a94ae/src/tiled/abstracttiletool.cpp#L145-L148

By finding usages of `setStatusInfo` you can find additional locations. You will also see that for objects tools, it includes the pixel coordinates as well. These are the locations that would need to be adjusted, such that in terms of the displayed coordinate, the origin is in the bottom-left (Y axis inverted).

In addition, I mentioned the coordinates displayed in the Properties view for a selected object. This happens here:

https://github.com/bjorn/tiled/blob/da9ba8f0c7138850b7c2b9b23e6544d0705a94ae/src/tiled/propertybrowser.cpp#L1572-L1573

Since we should not confuse the user by presenting two coordinate systems, these coordinates will need to be inverted as well. But internally the coordinates will stay like they are, so when the user changes these values, they need to be inverted again before applying them to the object (in `applyMapObjectValue`).

Finally (if I don't forget anything), the Objects view can optionally display object coordinates. This is done over here:

https://github.com/bjorn/tiled/blob/da9ba8f0c7138850b7c2b9b23e6544d0705a94ae/src/tiled/mapobjectmodel.cpp#L128-L133

Since there are so many places spread over the code where coordinates will need to be inverted, it will make sense to write a small helper class for this, which should check the user preferences (see `preferences.h`, which is where you'll add this option) and perform the conversion when needed.

**Comment by divitto** [comment]
Awesome i understand. I will get right on this as soon as possible. :)

**Comment by divitto** [comment]
@bjorn  I Believe i have completed the task. The only problem i am having is that if you place an object in standard mode and then switch to inverted coordinates (or vice-versa) the object you placed will not automaticly update it's coordinates in the PropertyBrowser, you have to move the object to make it update the coordinates in the PropertyBrowser. How do i get the PropertyBrowser to update it's display when the user clicks my checkBox in the preferences? I hope that is clear enough, I am still pretty new to the signals and slots system. It seems like i would want to create a signal whenever the checkBox is clicked that is connected to a slot that updates the PropertyBrowser but i can't seem to get it figured out. Otherwise everything looks great, any ideas?

**Comment by bjorn** [comment]
@divitto Sounds great! Please open a pull request so I could have a look at the implementation.

Indeed I expect you have added a boolean property to the `Preferences` class that determines whether the Y axis should be inverted. This property should have a signal emitted when it is changed, like most other preferences, and the `PropertyBrowser` should connect to this and make sure to update the properties in case it is displaying a `MapObject`. It is not the first signal from the preferences that the `PropertyBrowser` connects to: it will look pretty similar to its connection to the `Preferences::objectTypesChanged` signal (except that the slot should call `updateProperties` instead of `updateCustomProperties`).

**Comment by divitto** [comment]
@bjorn I created the pull request a few days ago but i just noticed the little red X next to my request. Did everything work okay?

**Comment by bjorn** [comment]
@divitto Let's discuss that on the pull request (#1897). :-)

**Comment by bjorn** [comment]
Given the issue I raised in https://github.com/mapeditor/tiled/pull/2040#issuecomment-450242395, I've removed the "good first issue" and "help wanted" labels. Issues with open design questions aren't very good first issues, and I'm not actively looking for help with this issue either.

**Comment by bryanedds** [comment]
At this point, I'm starting to think the correct thing to do is more than just shallow changes; an entire separate fork of Tiled needs to be created and maintained that flips Y at all levels - from UI down to data and data format. The industry has pretty much converged on positive Y being up in 2D and anything otherwise is just pure legacy. I think we need to at some point make an option for a clean break and for those who do, never look back.

Could we come up with an effort estimate for that in terms of an initial change list?
